### ISSUE TITLE
TF Distributed not using the full network bandwith

### ISSUE THREAD

**Issue by JianbangZ** [tensorflow/tensorflow#4498]
Hi,
I have a one ps, 4 worker setup (2 workers are in the same machine as ps), I found no matter what batch size I choose, the network usage is always around 1.3 Gbit/s, which leads to very slow performance with small batch size. However, I do have 40 Gbit network, and when I run iperf -s/c between these two machines, I can usually get over 30 Gbit/s speed. 

I am not sure if the issue is in TF implementation or my network setup. but since iperf proved, I doubt it's my infrastructure problem.  

Below is the configuration 
ps machine: Ubuntu 14.04, TF v0.10.0
worker machine: CentOS 7, TF v0.10.0 
Python 2.7
GPU is Tesla M40

**Comment by yaroslavvb** [comment]
Are you bottlenecked by CPU?

**Comment by yaroslavvb** [comment]
We are also seeing slowness an order of magnitude worse what's expected, and I tried to isolate the problem in a simple benchmark: https://gist.github.com/yaroslavvb/1124bb02a9fd4abce3d86caf2f950cb2

It launches 2 local tensorflow processes and tries to add 10MB vectors of in one process to variable stored in another one as fast as possible.

To run:
```
> python local_distributed_benchmark.py
....
Local rate:       6932.62 MB per second
Distributed rate: 161.63 MB per second
```

The rate of data tops out at 1.3 Gbps on my laptop, I wonder if it's a coincidence. Running on bigger machine doesn't give a speed increase, and CPU utilization stays under 20% . Same with latest nightly (GIT_VERSION='v0.10.0-1452-gdcba217-dirty')

**Comment by mrry** [comment]
Hi Yaroslav, I think your program is bottlenecked by this line:

```python
sess.run(add_op)
```

This ensures that the entire result is copied back from the worker to the master to the client, deserialized from a protobuf into Python and discarded. A simple change to the following makes the numbers more promising:

```python
sess.run(add_op.op)
```

* Before (with `sess.run(add_op)`):

   ```
Adding data in 10 MB chunks
Local rate:       3693.02 MB per second
Distributed rate: 115.94 MB per second
```

* After (with `sess.run(add_op.op)`):

   ```
Adding data in 10 MB chunks
Local rate:       10565.81 MB per second
Distributed rate: 6379.86 MB per second
```

Indeed, I'm almost certain that the placer will put the constant in your program  on the same device as the `assign_add()` (because it has only one consumer), so this is purely measuring dispatch overhead. Changing `update` to be a `tf.Variable`, so that it is pinned to the different device gives a more realistic number that falls in between the two cases above:

* With tensors pinned to different devices (and `sess.run(add_op.op)`):

   ```
Adding data in 10 MB chunks
Local rate:       10259.25 MB per second
Distributed rate: 832.56 MB per second
```

This throughput obviously still isn't as good as it could be, but I haven't tried very hard to saturate the "network" (e.g. by issuing parallel transfers). @JianbangZ, there are many different reasons why your code might not be saturating your 40GbE network. Being CPU-bound is the most likely reason, if your network is doing anything non-trivial. It's also possible that there's contention on the parameter server, or in some part of the TensorFlow runtime. Without more details about the code you are running, it's impossible to pinpoint the root cause, so if you can share more details, that'd be great!

**Comment by yaroslavvb** [comment]
@mrry Thanks for the quick response! Haven't realized that "tf.device" was ignored here.  I'm now getting 900 MB/s which is pretty good. So it looks like deserializing gRPC into Python is extremely slow (<1 Gbps regardless of number of cores) and should be avoided. @JianbangZ are you by chance fetching large tensors in your Python client?

**Comment by JianbangZ** [comment]
@mrry My code is a mimic of the inception distributed example, the code to control training op is the same as this https://github.com/tensorflow/models/blob/master/inception/inception/inception_distributed_train.py

Has the team seen data transfer slowness with this implementation?

**Comment by mrry** [comment]
@JianbangZ What's the duration of each step when running on a single process, and what's the size of the model that's being retrieved on each step? What batch size are you using? For Inception in the default configuration, I wouldn't expect it to saturate 40GbE, because the computation will be the bottleneck.

And just to confirm, are you running 0.10 or 0.10rc0 (because there were several performance fixes after the release candidate)?

**Comment by JianbangZ** [comment]
@mrry Thank you for the quick response. I am not training on GoogleNet, I am just using the method it uses to do distributed on my own network.
My net only has 3 conv + 2 fc + 1 softmax readout layers to handle the input image size of 99*99. The total parameter count is about 9 M, which means the parameter size is 36 MB, so 40 GbE is way enough.

My training speed is around 600 images/sec on single GPU/single process. Usually I use batch_size=128. with single machine 2 GPUs, so I can get 1200 images/sec. 
With distributed, no matter what batch size I try, it shows my network usage is around 1~1.3 Gbit/s, thus smaller batch size shows decreased performance. 

I am using the 0.10.0 pip binary.

To make this easy, I uploaded my code to github, although it can't run b/c it's tied my with my own dataset and couple of environment variables, you can still see all my code.
The script to kick off the task is https://github.com/JianbangZ/eva/blob/master/eva/applications/facial_analysis/distributed_train.sh
and the code to control most of the training logic is https://github.com/JianbangZ/eva/blob/master/eva/platform/base_distributed_train.py
I am sorry this file has a lot of commented out multi-GPU code.
the code to define my network architect is here (I am using my own CustomNet although I implemented other nets in the same file) https://github.com/JianbangZ/eva/blob/master/eva/applications/facial_analysis/cnn_model.py#L625

**Comment by drpngx** [comment]
The github code appears to have disappeared? Is this issue still current?

**Comment by aselle** [comment]
Automatically closing due to lack of recent activity, we will reopen if further information becomes available. Thanks!

**Comment by llhe** [comment]
#6116
